include/deal.II/base/tensor_product_polynomials_bubbles.h

```diff
@@ -176,7 +176,7 @@ template <int dim>
 inline unsigned int
 TensorProductPolynomialsBubbles<dim>::n() const
 {
-  return this->n_tensor_pols + dim;
+  return TensorProductPolynomials<dim>::n() + dim;
```

Comment:
Oh, this is awkward -- you are in essence overloading a member function from the base class but giving it a different meaning. This suggests that the "is-a" relationship of `TensorProductPolynomialsBubbles` and `TensorProductPolynomials` classes that is usually implied when you derive from a base class is not the case here, but that maybe a "has-a" relationship of using `TensorProductPolynomials` as a member variable would be more appropriate?

Reply:
Well the two commits marked as (broken) were originally going to try to replace all instances of `n_tensor_pols` in the derived class with `this->n()`; however, the number of polynomials for the derived class is actually `n_tensor_pols + dim`. I ran into a bunch of trouble with this and ended up reverting four files then redoing the broken commits.

At the moment, I have replaced all references to `n_tensor_pols` with the `TensorProductPolynomials::n()` and left the derived `n()` as `n_tensor_pols + dim`.

Should I instead just make a `n_tensor_pols` const member of this derived class to make this easier?

Reply:
I think giving the derived class a `n_tensor_pols` member might be an easier way to preserve the "is-a" relationship instead of turning it into a "has-a" relationship.

Reply:
Oh, but the base constructor is always called without giving it some notion of the additional functions, be them the +1 from const or +dim from bubbles. That's why this is a little tricky.

My goal with this PR is to preserve the original setup of this class without drastically changing it, since I'm only trying to add that base class.

Reply:
HM, but that doesn't sound right. What I'm saying is that if the number of polynomials is `TPP::n()+dim`, then someone has lied to the base class when calling the base class constructor. Nothing good can come of this: If you pass on object of this type via a reference-to-base, and the place where that argument is used calls `n()` on the object they have, then they will return the wrong number for the number of polynomials because this calls the `n()` function of the *base* class, not the *derived* class. 

The original sin here is to say that the `TPPBubbles` object *is a* TBB object, when in reality it is of course not: It's a polynomial space alright, but the shape functions are not constructed by taking tensor products. So the derivation may have seemed convenient at the time, but incorrect. That's what should be fixed first.

If that's a step too far for you, let me know and I can take care of it.

Reply:
I see. So perhaps we should have
```
template <int dim>
TensorProductPolynomialsBubbles : public ScalarPolynomialsBase<dim>
{
  ...
private:
  const TensorProductPolynomials<dim> tensor_polys;
}
```
and then we initialize this in the constructor with something like this?
```
TensorProductPolynomialsBubbles<dim>::TensorProductPolynomialsBubbles(const std::vector<Pol> &pols)
: ScalarPolynomialsBase<dim>(???, num_tensor_pols + dim)
, tensor_polys(pols)
```
I could handle that and make it a separate PR if that looks right to you.

Reply:
That particular implementation may also require `TensorProductPolynomials` and `TensorProductPolynomialsBubbles` to be friend classes?

Reply:
That's exactly what I was thinking of. This kind of derivation makes a lot more sense to me.

If you need something to be `friend`, mark it as such. It's often possible to work around this (and that also makes the code cleaner) but sometimes it just makes sense to do it with a `friend` declaration.

Reply:
Okay great! I was thinking of using a `friend` since `TensorProductPolynomials` has protected members `polynomials`, `index_map`, and `index_map_inverse`. This way I can rely on those instead of storing them a second time in `TensorProductPolynomialsBubbles` since it "has-a" `TensorProductPolynomials`.

Note to self: change those protect members to private members